acceptorwrapper/rate_limiting_conn.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) TFG Co. All Rights Reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package acceptorwrapper
+
+import (
+	"container/list"
+	"net"
+	"time"
+
+	"github.com/topfreegames/pitaya"
+	"github.com/topfreegames/pitaya/constants"
+	"github.com/topfreegames/pitaya/logger"
+	"github.com/topfreegames/pitaya/metrics"
+)
+
+// RateLimitingConn wraps net.Conn by applying rate limiting
+// and return empty if exceeded
+type RateLimitingConn struct {
+	net.Conn
+	limit        int
+	interval     time.Duration
+	times        list.List
+	forceDisable bool
+}
+
+// NewRateLimitingConn returns an initialized *RateLimiting
+func NewRateLimitingConn(
+	conn net.Conn,
+	limit int,
+	interval time.Duration,
+	forceDisable bool,
+) *RateLimitingConn {
+	r := &RateLimitingConn{
+		Conn:         conn,
+		limit:        limit,
+		interval:     interval,
+		forceDisable: forceDisable,
+	}
+
+	r.times.Init()
+
+	return r
+}
+
+func (r *RateLimitingConn) Read(b []byte) (n int, err error) {
+	if r.forceDisable {
+		return r.Conn.Read(b)
+	}
+
+	n, err = r.Conn.Read(b)
+	if err != nil {
+		return n, err
+	}
+
+	now := time.Now()
+	err = r.take(now)
+	if err != nil {
+		logger.Log.Errorf("Data=%s, Error=%s", b, err)
+		metrics.ReportExceededRateLimiting(pitaya.GetMetricsReporters())
+		return 0, nil
+	}
+
+	return n, err
+}
+
+// take saves the now as time taken or returns an error if
+// in the limit of rate limiting
+func (r *RateLimitingConn) take(now time.Time) error {
```

Comment:
nit: `take` is not a very descriptive name. I'd suggest something that alludes to rate limiting, maybe `shouldRateLimit() bool`?

Reply:
I agree about the name. 

Don't you think it's good to return the error here instead of only a bool?

Reply:
Either works, no strong opinion here. I took it out because it seems it only exists to be logged, in which case you could just write the message directly on `logger.Log.Errorf`.

Reply:
Fair enough. And `shouldRateLimit` makes more sense if returns a bool. I will change it.